frontends/p4/def_use.h

```diff
@@ -49,38 +49,44 @@ class StorageLocation : public IHasDbPrint {
     }
     cstring toString() const { return name; }
 
-    // all locations inside that represent valid bits
+    /// @returns All locations inside that represent valid bits.
     const LocationSet* getValidBits() const;
     virtual void addValidBits(LocationSet* result) const = 0;
-    // set of locations if we exclude all headers
+    /// @returns All locations inside if we exclude all headers.
     const LocationSet* removeHeaders() const;
     virtual void removeHeaders(LocationSet* result) const = 0;
-    // all locations inside that represent the last index of an array
+    /// @returns All locations inside that represent the 'lastIndex' of an array.
     const LocationSet* getLastIndexField() const;
     virtual void addLastIndexField(LocationSet* result) const = 0;
 };
 
-/* Represents a storage location with a simple type.
-   It could be either a scalar variable, or a field of a struct, etc. */
+/** Represents a storage location with a simple type or a tuple type.
+    It could be either a scalar variable, or a field of a struct, etc. */
 class BaseLocation : public StorageLocation {
  public:
+    // We can use this for tuples because tuples have no field accessors,
+    // so we treat them as monolithic objects.
     BaseLocation(const IR::Type* type, cstring name) :
             StorageLocation(type, name)
-    { BUG_CHECK(type->is<IR::Type_Base>() || type->is<IR::Type_Enum>()
-                || type->is<IR::Type_Var>() || type->is<IR::Type_Tuple>()
-                || type-is<IR::Type_Error>() || type->is<IR::Type_Var>(),
+    { BUG_CHECK(type->is<IR::Type_Bits>() || type->is<IR::Type_Enum>() ||
+                type->is<IR::Type_Boolean>() || type->is<IR::Type_Var>() ||
+                type->is<IR::Type_Tuple>() || type-is<IR::Type_Error>(),
```

Comment:
Isn't C++ beautiful? 'is' is a method on 'this', and this is just pointer arithmetic!
Sick, I will push a fix. Note that the bug has been there for a longer time.
And I was counting on the C++ type-checker...